compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIETileForMatmulLikeInnerLoop.cpp

```diff
@@ -0,0 +1,168 @@
+
+namespace mlir::iree_compiler::AMDAIE {
+
+namespace {
+
+class AMDAIETileForMatmulLikeInnerLoopPass
+    : public impl::AMDAIETileForMatmulLikeInnerLoopBase<
+          AMDAIETileForMatmulLikeInnerLoopPass> {
+  void getDependentDialects(DialectRegistry &registry) const final {
+    registry.insert<linalg::LinalgDialect, scf::SCFDialect>();
+  }
+
+  void runOnOperation() final;
+};
+
+class TileForMatmulPattern : public OpRewritePattern<linalg::GenericOp> {
+ public:
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    auto iteratorTypes = genericOp.getIteratorTypesArray();
+    auto numIterators = iteratorTypes.size();
+
+    if (numIterators < 4) {
+      return rewriter.notifyMatchFailure(
+          genericOp,
+          llvm::formatv("The number of iterators ({0}) is less than 4, so "
+                        "there are no outer dimensions to tile.",
+                        numIterators));
+    }
+
+    if (genericOp->getNumOperands() != 3) {
+      return rewriter.notifyMatchFailure(
+          genericOp, llvm::formatv("The number of operands ({0}) is not 3, so "
+                                   "this is not matmul-like.",
+                                   genericOp->getNumOperands()));
+    }
+
+    // Example: (1,1,4,1,2,3,4) has 5 dimensions from the first non-singleton.
+    auto getNumDimsFromFirstNonSingleton = [&](Value v) -> int {
+      auto shape = v.getType().cast<ShapedType>().getShape();
+      auto nDims = shape.size();
+      auto firstNonSingleton = [&]() -> int {
+        for (auto i : llvm::enumerate(shape)) {
+          if (i.value() > 1) {
+            return i.index();
+          }
+        }
+        return nDims;
+      }();
+      return nDims - firstNonSingleton;
+    };
+
+    // Don't transform to scf.for loops unless there is at least one
+    // non-singleton loop to construct (this ensures termination of the pass).
+    {
+      auto numDimsFromFirstNonSingleton0 =
+          getNumDimsFromFirstNonSingleton(genericOp->getOperand(0));
+      auto numDimsFromFirstNonSingleton1 =
+          getNumDimsFromFirstNonSingleton(genericOp->getOperand(1));
+      auto numDimsFromFirstNonSingleton2 =
+          getNumDimsFromFirstNonSingleton(genericOp->getOperand(2));
+
+      if (numDimsFromFirstNonSingleton0 < 3 &&
+          numDimsFromFirstNonSingleton1 < 3 &&
+          numDimsFromFirstNonSingleton2 < 3) {
+        return rewriter.notifyMatchFailure(
+            genericOp,
+            llvm::formatv(
+                "The number of non-singleton trailing dimensions is less than "
+                "3 for all operands ({0}, {1}, {2}), so there are no outer "
+                "dimensions to tile.",
+                numDimsFromFirstNonSingleton0, numDimsFromFirstNonSingleton1,
+                numDimsFromFirstNonSingleton2));
+      }
+    }
+
```

Comment:
If you can refactor the above `numDimsFromFirstNonSingleton` to a SmallVector, I think you can use llvm::any_of for the condition.